### PR TITLE
Refactoring of rotation search to improve performance

### PR DESCRIPTION
**Pull request by fsimkovic**
## Added
- `PdbStructure` to cache a pdb file only once but allow access of various functions
## Changed
- Functions converted to `MatthewsCoefficient` and `SolventContent` classes for multiple access
- `AmoreRotationSearch` refactored for big performance improvements

### CODE REVIEW

File: simbad/util/matthews_coef.py [hunk 1]
```diff
@@ -1,71 +1,40 @@
...
 
 
 class SolventContent(object):
     def __init__(self, cell, sg):
         self.crystal_symmetry = symmetry(unit_cell=cell, space_group_symbol=sg)
         self.dens_calc = density_calculator(self.crystal_symmetry)
 
-    def calculate(self, pdb):
-        return self.dens_calc.solvent_fraction(molecular_weight(pdb), 0.74) * 100
+    def calculate_form_file(self, pdb):
```
Comment (hlasimpk) [inline simbad/util/matthews_coef.py]: This may be fixed in a latter commit, but should be calculate_from_file

Comment (fsimkovic) [reply simbad/util/matthews_coef.py]: Good spot!

File: simbad/mr/__init__.py [hunk 2]
```diff
@@ -323,41 +324,46 @@ def submit_jobs(self, results, nproc=1, process_all=False, submit_qtype=None, su
...
+            ref_pdbout = os.path.join(ref_workdir,
+                                      '{0}_refinement_output.pdb'.format(result.pdb_code))
+
+            diff_mapout1 = os.path.join(ref_workdir,
+                                        '{0}_refmac_2fofcwt.map'.format(result.pdb_code))
+            diff_mapout2 = os.path.join(ref_workdir,
+                                        '{0}_refmac_fofcwt.map'.format(result.pdb_code))
+
+            pdb_struct = PdbStructure(result.dat_path)
+            mr_pdbin = os.path.join(self.output_dir, result.pdb_code + ".pdb")
```
Comment (hlasimpk) [inline simbad/mr/__init__.py]: This was a neat idea!

File: simbad/util/matthews_coef.py (hunk 1 above)
Comment (fsimkovic) [reply simbad/util/matthews_coef.py]: Sorted in 5e1c71d

File: simbad/rotsearch/amore_search.py [hunk 3]
```diff
@@ -40,12 +36,13 @@ class AmoreRotationSearch(object):
...
         The maximum number of results to keep [default: 20]
+    search_results : list
+        The search results
 
     Examples
     --------
     >>> from simbad.rotsearch.amore_search import AmoreRotationSearch
     >>> rotation_search = AmoreRotationSearch('<amore_exe>', '<mtz>', '<work_dir>', '<max_to_keep>')
-    >>> rotation_search.sortfun()
     >>> rotation_search.run_pdb(
```
Comment (hlasimpk) [inline simbad/rotsearch/amore_search.py]: This example needs updating to reflect the name change from run_pdb to run

File: simbad/rotsearch/amore_search.py [hunk 4]
```diff
@@ -60,298 +57,25 @@ class AmoreRotationSearch(object):
...
-        stdin = os.linesep.join([
-            "TITLE: Produce table for MODEL FRAGMENT",
-            "TABFUN",
-            "CRYSTAL {0} {1} {2} {3} {4} {5} ORTH 1",
-            "MODEL 1 BTARGET 23.5",
-            "SAMPLE 1 RESO 2.5 SHANN 2.5 SCALE 4.0",
-        ])
-        stdin = stdin.format(x, y, z, a, b, c)
-        return cmd, stdin
+    def run_pdb(self, *args, **kwargs):
```
Comment (hlasimpk) [inline simbad/rotsearch/amore_search.py]: My mistake above, this is still here. Do we need this?

Comment (fsimkovic) [reply simbad/rotsearch/amore_search.py]: I left it in to not break it, but probs going to remove it to address this and previous comment

Comment (hlasimpk) [reply simbad/rotsearch/amore_search.py]: From looking at comand_line/__init__, every where it was used has been replaced by run and seeing as we've not release this yet we don't have to worry about continuity just yet.

Comment (fsimkovic) [reply simbad/rotsearch/amore_search.py]: Addressed in c01b3ab

File: simbad/rotsearch/amore_search.py (hunk 3 above)
Comment (fsimkovic) [reply simbad/rotsearch/amore_search.py]: Addressed in c01b3ab

File: simbad/rotsearch/amore_search.py [hunk 5]
```diff
@@ -378,169 +102,127 @@ def run_pdb(self, models_dir, mr_model_dir, nproc=2, shres=3.0, pklim=0.5, npic=
...
-                except ValueError:
-                    logger.critical("Error calculating solvent content for %s",
-                                    name)
-                    continue
+        tmp_dir = os.path.join(self.work_dir, "tmp")
+        if not os.path.isdir(tmp_dir):
+            os.makedirs(tmp_dir)
 
+        hklpck0 = self._generate_hklpck0()
+	template_hklpck1 = os.path.join(tmp_dir, "{}.hkl")
```
Comment (hlasimpk) [inline simbad/rotsearch/amore_search.py]: Can you double check the indexing here, it seems to be one space off

Comment (fsimkovic) [reply simbad/rotsearch/amore_search.py]: Addressed in 35f4195

File: simbad/rotsearch/amore_search.py [hunk 6]
```diff
@@ -560,3 +243,89 @@ def summarize(self, csv_file):
         ]
         summarize_result(self.search_results,
                          csv_file=csv_file, columns=columns)
+
+    def _generate_hklpck0(self):
+        logger.info("Preparing files for AMORE rotation function")
+        f, sigf, _, _, _, _, _ = simbad.util.mtz_util.get_labels(self.mtz)
+        stdin = self.sortfun_stdin_template.format(f=f, sigf=sigf)
```
Comment (hlasimpk) [inline simbad/rotsearch/amore_search.py]: Out of curiosity, where is the template specified?

Comment (fsimkovic) [reply simbad/rotsearch/amore_search.py]: ~20 lines down

Comment (hlasimpk) [reply simbad/rotsearch/amore_search.py]: cheers, couldn't see it for the life of me just then.

Comment (fsimkovic) [reply simbad/rotsearch/amore_search.py]: I moved all `property` attributes to the bottom for clarity of the functions near the top.

File: simbad/rotsearch/amore_search.py [hunk 7]
```diff
@@ -378,169 +102,127 @@ def run_pdb(self, models_dir, mr_model_dir, nproc=2, shres=3.0, pklim=0.5, npic=
...
+                pdb_model = template_model.format(name)
+                table1 = template_table1.format(name)
+                hklpck1 = template_hklpck1.format(name)
+                clmn0 = template_clmn0.format(name)
+                clmn1 = template_clmn1.format(name)
+                mapout = template_mapout.format(name)
+
+                conv_cmd = ["ccp4-python", "-c",
+                            "\"import simbad.db; simbad.db.convert_dat_to_pdb('{}', '{}')\""]
+                conv_cmd[-1] = conv_cmd[-1].format(dat_model, pdb_model)
```
Comment (hlasimpk) [inline simbad/rotsearch/amore_search.py]: This comment is just semantics but...

Would it be easier to read if this was reversed to:

conv_cmd =  "\"import simbad.db; simbad.db.convert_dat_to_pdb('{}', '{}')\"".format(dat_model, pdb_model)
conv_cmd = ["ccp4-python", "-c", conv_cmd]

or is there a benefit to creating the list first?

Comment (fsimkovic) [reply simbad/rotsearch/amore_search.py]: Thanks ... that was a late "hack" that's clearly gone wrong 🦆   Addressed in 4caace2

**Pull request merged**